locksmith/src/websub/jobs/keys.ts

```diff
@@ -97,6 +116,284 @@ async function notifyHooksOfAllUnprocessedKeys(hooks: Hook[], network: number) {
   }
 }
 
+const getRecipient = async (lockAddress: string, ownerAddress: string) => {
+  const metadata = await userMetadataOperations.getMetadata(
+    lockAddress,
+    ownerAddress,
+    true // include protected
+  )
+
+  const protectedData = Normalizer.toLowerCaseKeys({
+    ...(metadata?.userMetadata?.protected ?? {}),
+  })
+
+  return protectedData?.email as string
+}
+
+const getMembershipStatus = async ({
+  key,
+  tokenAddress,
+  network,
+  lockAddress,
+  tokenId,
+  owner,
+}: {
+  key: any
+  tokenAddress: string
+  network: number
+  lockAddress: string
+  tokenId: string
+  owner: string
+}) => {
+  const isERC20 = tokenAddress && tokenAddress !== ethers.constants.AddressZero
+
+  const isRenewable =
+    Number(key?.lock?.version) >= 11 && key?.expiration !== MAX_UINT && isERC20
+
+  const subscriptions =
+    await subscriptionOperations.getSubscriptionsForLockByOwner({
+      tokenId,
+      lockAddress,
+      owner,
+      network,
+    })
+
+  let currency = ''
+  let isAutoRenewable = false
+  let isRenewableIfReApproved = false
+
+  // TODO: need new version of the lock contract to get the value easily
+  const isRenewableIfRePurchased = false
+
+  // get subscription and check for renews
+  if (subscriptions?.length) {
+    const [subscription] = subscriptions ?? []
+
+    if (subscription) {
+      const possible = ethers.BigNumber.from(subscription.possibleRenewals)
+      const approved = ethers.BigNumber.from(subscription.approvedRenewals)
+
+      isAutoRenewable = approved.gte(0) && possible.gte(0)
+
+      isRenewableIfReApproved = approved.lte(0)
+
+      currency = subscription.balance.symbol
+    }
+  }
+
+  return {
+    currency,
+    isRenewable,
+    isAutoRenewable,
+    isRenewableIfReApproved,
+    isRenewableIfRePurchased,
+  }
+}
+
+export async function notifyKeyExpiration() {
+  const now = new Date()
+
+  const end = new Date(now.getTime())
+  end.setDate(now.getDate() + 1)
+
+  await Promise.allSettled(
+    // get expiring keys for every network
+    Object.keys(networks).map(async (networkId: string) => {
+      const subgraph = new SubgraphService()
+      // get keys from now to 24h in the future
+      const keys = await subgraph.keys(
+        {
+          first: 1000, //  TODO: handle more than 1000 keys
+          where: {
+            expiration_gt: now.getTime().toString(),
+            expiration_lt: end.getTime().toString(),
+          },
+        },
+        {
+          networks: [Number(networkId)],
+        }
+      )
+
+      logger.info(
+        `keys expiring for ${networks[networkId].name}: ${keys?.length}`
+      )
+
+      // split keys into groups to send email by group with some delay
+      const chunks = arrayToChunks(keys, KEYS_CHUNKS_SIZE)
+      await Promise.allSettled(
+        Object.values(chunks).map(async (keys, index) => {
+          const timeout = DELAY * (index + 1) // increase delay for each chunks
```

Comment:
Let's create a subfunction here.

Reply:
Everytime you have something that's `Promise.allSettled` you should just pass a single function rather than a block. This clarifies your code a lot.